src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs

```diff
@@ -802,126 +1026,114 @@ private void AggregateRenderPackages(PackageAggregationParams parameters)
+                id = ((rp.RequiresPerVertexColoration || rp.Colors != null) ? rp.Description : baseId) + ":mesh";
 
-                    meshSet.Colors.AddRange(rp.IsSelected ? Enumerable.Repeat(selectionColor,m.Positions.Count): m.Colors);
-                    meshSet.Normals.AddRange(m.Normals);
-                    meshSet.TextureCoordinates.AddRange(m.TextureCoordinates);
-                    meshSet.Indices.AddRange(m.Indices.Select(i => i + idxCount));
+                DynamoGeometryModel3D meshGeometry3D;
 
-                    if (rp.DisplayLabels)
+                if (geometryDictionary.ContainsKey(id))
+                {
+                    meshGeometry3D = geometryDictionary[id] as DynamoGeometryModel3D;
+                }
+                else
+                {
+                    meshGeometry3D = new DynamoGeometryModel3D()
                     {
-                        var pt = meshSet.Positions[idxCount];
-                        parameters.Text.TextInfo.Add(new TextInfo(HelixRenderPackage.CleanTag(rp.Description), new Vector3(pt.X + 0.025f, pt.Y + 0.025f, pt.Z + 0.025f)));
+                        Geometry = HelixRenderPackage.InitMeshGeometry(),
+                        Transform = Model1Transform,
+                        Material = WhiteMaterial,
+                        IsHitTestVisible = false,
+                        RequiresPerVertexColoration = rp.RequiresPerVertexColoration,
+                        IsSelected = rp.IsSelected,
+                    };
+
+                    if (rp.Colors != null)
+                    {
+                        var pf = PixelFormats.Bgra32;
+                        var stride = (rp.ColorsStride / 4 * pf.BitsPerPixel + 7) / 8;
+                        var diffMap = BitmapSource.Create(rp.ColorsStride/4, rp.ColorsStride/4, 96.0, 96.0, pf, null, rp.Colors.ToArray(), stride);
+                        var diffMat = new PhongMaterial
+                        {
+                            Name = "White",
+                            AmbientColor = PhongMaterials.ToColor(0.1, 0.1, 0.1, 1.0),
+                            DiffuseColor = materialColor,
+                            SpecularColor = PhongMaterials.ToColor(0.0225, 0.0225, 0.0225, 1.0),
+                            EmissiveColor = PhongMaterials.ToColor(0.0, 0.0, 0.0, 1.0),
+                            SpecularShininess = 12.8f,
+                            DiffuseMap = diffMap
+                        };
+                        meshGeometry3D.Material = diffMat;
                     }
+                    ((MaterialGeometryModel3D) meshGeometry3D).SelectionColor = selectionColor; 
+                    geometryDictionary.Add(id, meshGeometry3D);
                 }
-            }
-        }
+                var meshSet = meshGeometry3D.Geometry as MeshGeometry3D;
+                var idxCount = meshSet.Positions.Count;
 
-        private void Detach(bool detachPersistentObjects)
```

Comment:
VisualizationManager_WorkspaceOpenedClosedHandled this is called when a workspace is opened or closed. Here, all the geometry (except the grid and axes) will be detached and removed from the dictionary.